src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.fetch;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.LeafReaderContext;
+import org.opensearch.common.document.DocumentField;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.common.bytes.BytesReference;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.fetch.FetchContext;
+import org.opensearch.search.fetch.FetchSubPhase;
+import org.opensearch.search.fetch.FetchSubPhaseProcessor;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.BYTES_PER_KILOBYTES;
+
+
+public class KNNFetchSubPhase implements FetchSubPhase {
```

Comment:
why do we need another subphase? 

I see in Opensearch there is already a FetchDocValuesSubphase here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/fetch/subphase/FetchDocValuesPhase.java#L55

Reply:
> why do we need another subphase?

@navneet1v  this subPhase is used to replace the '_source'. I will fulfill the docValues fields into '_source' and this can make reindex works with knn_fields excludes.

Reply:
So just for fetching vector fields from Doc Values we don't need this right?

Reply:
> So just for fetching vector fields from Doc Values we don't need this right?

yes.  

`KNNFetchSubPhase` class add a process in fetch phase, and it can fulfill the `_source` with docValues fields response. and this way something like synthetic source but need explicit add value from search body like `docvalue_fields`. so it must combine with `KNNVectorDVLeafFieldData`

Reply:
> Make sense. I want to understand how we are ensuring that this subphase doesn't run during normal queries?

I had a check with code, only for KNN field and not in the map :https://github.com/opensearch-project/k-NN/blob/dbaf420a2d8b2dd54411dcdced1c1e0ea50bc83c/src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java#L69